nx/source/runtime/devices/usb_comms.c

```diff
@@ -25,69 +26,135 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
+                // Upgrade packet size to 512
+                device_descriptor.bMaxPacketSize0 = 0x09;
+                if (R_SUCCEEDED(rc)) rc = usbDsSetUsbDeviceDescriptor(UsbDeviceSpeed_Super, &device_descriptor);
+                
+                // Define Binary Object Store
+                u8 bos[0x16] = {
+                    0x05, // .bLength
+                    USB_DT_BOS, // .bDescriptorType
+                    0x16, 0x00, // .wTotalLength
+                    0x02, // .bNumDeviceCaps
+                    
+                    // USB 2.0
+                    0x07, // .bLength
+                    USB_DT_DEVICE_CAPABILITY, // .bDescriptorType
+                    0x02, // .bDevCapabilityType
+                    0x02, 0x00, 0x00, 0x00, // dev_capability_data
+                    
+                    // USB 3.0
+                    0x0A, // .bLength
+                    USB_DT_DEVICE_CAPABILITY, // .bDescriptorType
+                    0x03, // .bDevCapabilityType
+                    0x00, 0x0E, 0x00, 0x03, 0x00, 0x00, 0x00
+                };
+                if (R_SUCCEEDED(rc)) rc = usbDsSetBinaryObjectStore(bos, sizeof(bos));
+            }
+            
+            if (R_SUCCEEDED(rc)) {
+                for (u32 i = 0; i < num_interfaces; i++) {
+                    usbCommsInterface *intf = &g_usbCommsInterfaces[i];
+                    rwlockWriteLock(&intf->lock);
+                    rwlockWriteLock(&intf->lock_in);
+                    rwlockWriteLock(&intf->lock_out);
+                    rc = _usbCommsInterfaceInit(i, infos + i);
+                    rwlockWriteUnlock(&intf->lock_out);
+                    rwlockWriteUnlock(&intf->lock_in);
+                    rwlockWriteUnlock(&intf->lock);
+                    if (R_FAILED(rc)) {
+                        break;
+                    }
+                }
+            }
+        }
+        
+        if (R_SUCCEEDED(rc) && kernelAbove500()) {
+            rc = usbDsEnable();
+        }
+        
+        if (R_FAILED(rc)) {
+            usbCommsExit();
         }
-
-        if (!found) rc = MAKERESULT(Module_Libnx, LibnxError_OutOfMemory);
-    }
-
-    if (R_SUCCEEDED(rc)) {
-        rwlockWriteLock(&inter->lock);
-        rwlockWriteLock(&inter->lock_in);
-        rwlockWriteLock(&inter->lock_out);
-        rc = _usbCommsInterfaceInit(inter, bInterfaceClass, bInterfaceSubClass, bInterfaceProtocol);
-        rwlockWriteUnlock(&inter->lock_out);
-        rwlockWriteUnlock(&inter->lock_in);
-        rwlockWriteUnlock(&inter->lock);
-    }
-
-    if (R_FAILED(rc)) {
-        usbCommsExit();
     }
-
-    if (R_SUCCEEDED(rc)) g_usbCommsInitialized=true;
-
-    if (R_SUCCEEDED(rc) && interface) *interface = i;
+    
+    if (R_SUCCEEDED(rc)) g_usbCommsInitialized = true;
 
     rwlockWriteUnlock(&g_usbCommsLock);
-
     return rc;
 }
 
-Result usbCommsInitialize(void)
+Result usbCommsInitialize(const UsbCommsInterfaceInfo *info)
```

Comment:
Can you change this to void and pass a hard-coded struct to usbCommsInitializeEx instead?